backend/src/data_entry/api.rs

```diff
@@ -313,6 +316,79 @@ async fn polling_station_data_entry_finalise(
     Ok(())
 }
 
+/// Resolve data entry differences by providing a `ResolveAction`
+#[utoipa::path(
+    post,
+    path = "/api/polling_stations/{polling_station_id}/data_entries/resolve",
+    request_body = ResolveAction,
+    responses(
+        (status = 200, description = "Differences resolved successfully"),
+        (status = 401, description = "Unauthorized", body = ErrorResponse),
+        (status = 404, description = "Not found", body = ErrorResponse),
+        (status = 409, description = "Request cannot be completed", body = ErrorResponse),
+        (status = 422, description = "JSON error or invalid data (Unprocessable Content)", body = ErrorResponse),
+        (status = 500, description = "Internal server error", body = ErrorResponse),
+    ),
+    params(
+        ("polling_station_id" = u32, description = "Polling station database id"),
+    ),
+)]
+async fn polling_station_data_entry_resolve(
+    _user: Coordinator,
+    State(polling_station_data_entries): State<PollingStationDataEntries>,
+    Path(polling_station_id): Path<u32>,
+    audit_service: AuditService,
+    resolve_action: ResolveAction,
+) -> Result<Json<PollingStationDataEntry>, APIError> {
+    let state = polling_station_data_entries
+        .get_or_default(polling_station_id)
+        .await?;
+
+    let data_entry = match resolve_action {
+        ResolveAction::KeepFirstEntry => {
+            let new_state = state.keep_first_entry()?;
+            let data_entry = polling_station_data_entries
+                .upsert(polling_station_id, &new_state)
+                .await?;
+            audit_service
+                .log(
+                    &AuditEvent::DataEntryKeptFirst(data_entry.clone().into()),
+                    None,
+                )
+                .await?;
+            data_entry
+        }
+        ResolveAction::KeepSecondEntry => {
+            let new_state = state.keep_second_entry()?;
+            let data_entry = polling_station_data_entries
+                .upsert(polling_station_id, &new_state)
+                .await?;
+            audit_service
+                .log(
+                    &AuditEvent::DataEntryKeptSecond(data_entry.clone().into()),
+                    None,
+                )
+                .await?;
+            data_entry
+        }
+        ResolveAction::DiscardBothEntries => {
+            let new_state = state.delete_entries()?;
+            let data_entry = polling_station_data_entries
+                .upsert(polling_station_id, &new_state)
+                .await?;
+            audit_service
+                .log(
+                    &AuditEvent::DataEntryDiscardedBoth(data_entry.clone().into()),
+                    None,
+                )
+                .await?;
+            data_entry
+        }
+    };
```

Comment:
This has a lot of duplicated code and can be made more compact, e.g. (untested!!):

```suggestion
    let (status, audit_event) = match resolve_action {
        ResolveAction::KeepFirstEntry =>
            (DataEntryStatus::keep_first_entry, AuditEvent::DataEntryKeptFirst),
        ResolveAction::KeepSecondEntry =>
            (DataEntryStatus::keep_second_entry, AuditEvent::DataEntryKeptSecond),
        ResolveAction::DiscardBothEntries =>
            (DataEntryStatus::delete_entries, AuditEvent::DataEntryDiscardedBoth),
    };

    let new_state = status(&state)?;
    let data_entry = polling_station_data_entries
        .upsert(polling_station_id, &new_state)
        .await?;
    audit_service
        .log(&audit_event(data_entry.clone().into()), None)
        .await?;
```

Reply:
I hadn't thought about a solution like this, I'll try it out. Thanks!

Reply:
I've reduced duplication in [fc2f891](https://github.com/kiesraad/abacus/pull/1320/commits/fc2f8911a7c800864857f975c5ce814abc53ffe3). Your solution didn't work because we can't return just enum variants from the match statements. It needs to contain the data that is returned by the upsert. I've refrained from returning a function pointer as I think just calling the transition function is a little more readable.

Thanks for the suggestion. The endpoint looks way cleaner now!

Reply:
Looks good indeed!